src/opers.c

```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
-                (Int)TNAM_OBJ(arg1), 0L, 
-                "you can replace the first argument <arg1> via 'return <arg1>;'");
-      }
-    type1 = FLAGS_FILT( arg1 );
-    type2 = TYPE_OBJ_FEO( arg2 );
-    id2 = ID_TYPE( type2 );
-    type3 = TYPE_OBJ_FEO( arg3 );
-    id3 = ID_TYPE( type3 );
-
-    /* try to find an applicable method in the cache                       */
-    prec = INTOBJ_INT(-1);
-
-    do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 3 ) );
-
-      /* Up to CACHE_SIZE methods might be in the cache */
-      if (prec < INTOBJ_INT(CACHE_SIZE))
-        {
-          for (i = 0;  i < 5*CACHE_SIZE; i+= 5) {
-            if (  cache[i+1] == prec && cache[i+2] == type1
-                  && cache[i+3] == id2 && cache[i+4] == id3 ) {
-              method = cache[i];
-#ifdef COUNT_OPERS
-              OperationHit++;
-#endif
-              break;
-            }
-          }
-        }
-      
-      /* otherwise try to find one in the list of methods                    */
-      if (!method)
-        {
-          if (prec == INTOBJ_INT(0))
-            method = CALL_4ARGS( Constructor3Args, oper, type1, type2, type3 );
-          else
-            method = CALL_5ARGS( NextConstructor3Args, oper, prec, type1, type2, type3 );
-          
-          while (method == Fail)
-            {
-              Obj arglist[3];
-              arglist[0] = arg1;
-              arglist[1] = arg2;
-              arglist[2] = arg3;
-              method = CallHandleMethodNotFound(oper, 3, arglist, 0, 1, prec);
-            }
-          
-          /* update the cache */
-          if (method && prec < INTOBJ_INT(CACHE_SIZE))
-            {
-              Bag cacheBag = GET_METHOD_CACHE( oper, 3 );
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[5*STATE(CacheIndex)] = method;
-              cache[5*STATE(CacheIndex)+1] = prec;
-              cache[5*STATE(CacheIndex)+2] = type1;
-              cache[5*STATE(CacheIndex)+3] = id2;
-              cache[5*STATE(CacheIndex)+4] = id3;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
-            }
-#ifdef COUNT_OPERS
-          OperationMiss++;
-#endif
-        }
-      if ( !method )  {
-        ErrorQuit( "no method returned", 0L, 0L );
-      }
-      
-      /* call this method                                                    */
-      res = CALL_3ARGS( method, arg1, arg2, arg3 );
-    }
-    while (res == TRY_NEXT_METHOD );
-
-    /* return the result                                                   */
-    return res;
-}
-
-
-/****************************************************************************
-**
-**  DoConstructor4Args( <oper>, <a1>, <a2>, <a3>, <a4> )
-*/
-Obj DoConstructor4Args (
-    Obj      
```

Comment:
mark these as `static`? That may help the optimizer, too.

Reply:
They are used elsewhere in the kernel quite a lot. It saves one memory access compared to CALL_4ARGS which is probably not actually a good choice, but making these static now seems like a decision that doesn't belong in this PR.

Reply:
They are used elsewhere? I am confused, the only file in which the string `DoConstructor4Args` occurs is `opers.c` (of course other files use it indirectly via function pointers).
Anyway, I don't mind either way :)

Reply:
You're right of course. DoOperation<n>Args for <n> <=3 are used elsewhere though. I'd be inclined to think they should all be static or none of them.